cmd/client.go

```diff
@@ -0,0 +1,151 @@
+//		- no probe extension in probe config
+//			-> The probe proto does not have an extension registered.
+//		- no probes registered for the extension: [extension code]
+//			-> The proto has a probe extension, but there is no probe extension registered matching that extension code.
+//		- proto.GetExtension() error: extension missing, invalid, incomplete or proto cannot be extended.
+//			-> The proto.SetExtension() method must be called and supplied with valid data.
+//		- unregistered user defined probe: [probe name]
+//			-> The probe type has not been registered as a user defined probe.
+//		- unknown probe type: [probe type]
+//			-> The probe type does not match any expected types (including user-defined and extension).
+//		- probe.Init() error: an error occurred when the probe, created from the config passsed, was initialised.
+//			-> There was a problem in the Init() function of the probe type passed.
+//		- options.BuildProbeOptions() error: an error occurred when building the options for this probe from the config supplied.
+//			-> The options could not be built from the config supplied.
+func (client *CloudproberClient) addProbeFromConfig(ctx context.Context, probePb *configpb.ProbeDef) error {
+	// No mutex locking handled here as this is a private method called by
+	// a public method which handles locking.
+	_, err := client.client.AddProbe(ctx, &proberpb.AddProbeRequest{ProbeConfig: probePb})
+	return err
+}
+
+// RegisterAndAddProbe registers a probe type and adds the probe to Cloudprober to be run.
+// Parameters:
+// - probePb: - This probe config must be unmarshalled before being passed as an argument.
+// Returns:
+// - error:
+//	      - See addProbeFromConfig errors.
+func (client *CloudproberClient) RegisterAndAddProbe(extensionNumber int, ctx context.Context, probePb *configpb.ProbeDef, hermesProbeToAdd cpprobes.Probe) error {
+	client.clientMux.Lock()
+	defer client.clientMux.Unlock()
+
+	cpprobes.RegisterProbeType(extensionNumber, func() cpprobes.Probe { return hermesProbeToAdd })
+
+	// Only one extension type per proto config can be registered at any given time.
+	// If more are registered, Cloudprober will throw an error.
+	return client.addProbeFromConfig(ctx, probePb)
+}
+
+// RemoveProbe removes a probe from Cloudprober, given the probe name (located in the probe config).
+// An error will be returned if there is no active probe with this probe name.
+// Parameters:
+// - probeName: This must be the name of an active probe in Cloudprober.
+// Returns:
+// - error:
+//	   - Code: 3, InvalidArgument: probeName is an empty string
+//	   - Code: 5,  NotFound: cannot find a probe matching this probe name
+func (client *CloudproberClient) RemoveProbe(probeName string) error {
+	client.clientMux.Lock()
+	defer client.clientMux.Unlock()
+
+	_, err := client.client.RemoveProbe(context.Background(), &proberpb.RemoveProbeRequest{ProbeName: &probeName})
```

Comment:
The background context should only be used in main() and in tests.
Add a `ctx` argument to the signature.
Here and elsewhere.

Reply:
I have added a `ctx` argument to all of these function signatures.